File: sdks/csharp/Protocol/Registration.cs
namespace Semacore.Protocol;

/// <summary>事件订阅句柄：<see cref="Unregister"/> 取消订阅（≙ Java protocol/Registration）。</summary>
public sealed class Registration : IDisposable
{
    private Action? _unregister;

    public Registration(Action unregister) => _unregister = unregister;

    public void Unregister() => Interlocked.Exchange(ref _unregister, null)?.Invoke();

    public void Dispose() => Unregister();
}

File: sdks/java/src/main/java/semacore/protocol/Registration.java
package semacore.protocol;

/** 事件订阅句柄：{@link #unregister()} 取消订阅。 */
@FunctionalInterface
public interface Registration extends AutoCloseable {

    void unregister();

    @Override
    default void close() {
        unregister();
    }
}
